### DISCUSSION TITLE
High complexity in the '_add_missing_reverse_references' method.

### DISCUSSION THREAD

**Discussion by xydan83** [FreeOpcUa/opcua-asyncio#736] [Q&A]
Hi everybody!:)
When trying to import with a large number of nodes (from 1 thousand to 10 thousand), I noticed that the import takes a very long time. I started profiling and it turned out there is a method that has a minimum complexity of O(n ^ 3).
Was added in this [commit](https://github.com/FreeOpcUa/opcua-asyncio/commit/b0c3d576a5d2cd6603eccf3b8a7411d5b9eb984e)

Import of 1 thousand nodes: 0:00:52
Import of 10 thousand nodes: 3:46:00
Profiling indicated that more time was spent on **_add_missing_reverse_references()** method in the **xmlimporter.py**.

Is it possible to search for and add missing reverse references in a different way?

**Comment by swamper123** [comment]
hey ho,

well we have different problems here. One is, that Python isn't performing well with XML (which is one part of not so good performance).
The other one is the problem, I mentioned in the [comment to the commit](https://github.com/FreeOpcUa/opcua-asyncio/commit/b0c3d576a5d2cd6603eccf3b8a7411d5b9eb984e#commitcomment-60865801) itself.

Since a new node has different reference types to different existing nodes and the possibility of a faulty node exists (caused by wrong programming or faulty xml file) you have to check that...and touch every single one to be sure. I couldn't figure out a more performing way. 

What exists (but is more hidden in backround and atm I am not a friend of it) is to[ pickle the whole aspace, when it got everything you need, and dump or load it.](https://github.com/FreeOpcUa/opcua-asyncio/blob/68b50bbf458d892965c80dfd44590defd2ec4a98/asyncua/server/address_space.py#L605). But you would still need to run this process atleast once, so that everything would be still alright with the references.

**Reply by swamper123** [reply]
In case of better ideas, comments are welcomed and PRs even more. :^)

**Reply by xydan83** [reply]
As far as I understand, the dump can only be done on the server, which means that you need to already have a loaded structure on the server.
Can the client collect the address space inside himself to make a dump?
This is necessary, for example, to make a dump from another OPC server (not only opcua-asyncio).

**Reply by swamper123** [reply]
> As far as I understand, the dump can only be done on the server, which means that you need to already have a loaded structure on the server.

Correct.

>  Can the client collect the address space inside himself to make a dump?

We haven't got a feature for that. 
If I understood you idea right, this wouldn't be a good idea either, since you had to browse all Nodes from a server and then pickle the structure somehow. It doesn't sound impossible, but I see different problems with this plan.

**Reply by xydan83** [reply]
> We haven't got a feature for that. If I understood you idea right, this wouldn't be a good idea either, since you had to browse all Nodes from a server and then pickle the structure somehow. It doesn't sound impossible, but I see different problems with this plan.

Yes, in fact, you would have to make some kind of "internal server", but without interfaces to the outside, which would only store the address space. But for this you have to do browsing and read each node in order to add it to the server. Essentially the same "import". Perhaps it really doesn't make sense. 🤔

**Reply by xydan83** [reply]
The idea came to try to parallelize the search for each node by processes. There is a general loop that iterates over all the nodes and from it is already looking for children and their links, in which it looks for a parent, and if it does not find it, it adds a reference reverse. I think we can try to separate this process for processing each such node into a separate python process. Let's say we find out the number of cores in the processor and create an equal number of processes. It seems that there is no dependence between the nodes themselves. In theory, the import speed will increase by the number of cores. Maybe the decision is dubious, but so far I can not think of anything better.
I do not consider threads, since CPython blocks them (GIL).
![image](https://user-images.githubusercontent.com/86729310/143546096-985516e2-c462-4a50-aec7-32ccb13bc890.png)

**Comment by xydan83** [comment]
Another option is to single out one process and add missing reverse references in the background, but I don't quite understand if it is possible to do this?

**Reply by swamper123** [reply]
I agree, that this may speed up things, but it may cause other issues. Imagine you want to add 3 XML files which depend on each other (like Companion specifications). Example: [MachineTool](https://github.com/OPCFoundation/UA-Nodeset/blob/f71b3f411d5cb16097c3ae0c744f67ad45535ffb/MachineTool/Opc.Ua.MachineTool.NodeSet2.xml#L39) needs DI, Machinery and IA (and UA in Version 1.04.7 upwards, but this should be the server base). If you spwan 3 Processes now and IA can't find it's Nodes/References you got dangling or none existing Nodes in the aspace.
You would have to wait that each process ends succesfully or create a mechanism around it like a worker queue.
Even than, you may want to add some Nodes hardcoded or do stuff with hardcoded nodes which may only exist in these XML files. It still is a tricky situation.

To break the O(n^3) thing, we would have to eliminate the for loops (which cause this trouble, since 3 are nested). What could work would be some list comprehension (which would be more complex to read) magic. In InternalServer, the aspace has a dict attribute self._nodes, which contains each node on the server. Perhaps this might be a way to go.

**Reply by xydan83** [reply]
It was possible to speed up the processing a little by using a dictionary instead of a list on the last loop. It may have been possible to reduce the complexity from O (n ^ 3) to O (n ^ 2) at best.

I checked it on a test with 10 thousand nodes.
**Was: 3:46:00.204322
Now: 1:30:33.742987**

```python
    async def _add_missing_reverse_references(self, new_nodes):
        __unidirectional_types = {ua.ObjectIds.GuardVariableType, ua.ObjectIds.HasGuard,
                                  ua.ObjectIds.TransitionVariableType, ua.ObjectIds.StateMachineType,
                                  ua.ObjectIds.StateVariableType, ua.ObjectIds.TwoStateVariableType,
                                  ua.ObjectIds.StateType, ua.ObjectIds.TransitionType,
                                  ua.ObjectIds.FiniteTransitionVariableType, ua.ObjectIds.HasInterface}
        dangling_refs_to_missing_nodes = set()
        for new_node_id in new_nodes:
            new_n = self.server.get_node(new_node_id)
            new_n_refs = await new_n.get_references()
            if len(new_n_refs) == 0:
                _logger.warning("Node %s has no references, so it does not exist in Server!", new_node_id)
                continue
            for ref in new_n_refs:
                if ref.ReferenceTypeId not in __unidirectional_types:
                    n = self.server.get_node(ref.NodeId)
                    n_refs = {ref.NodeId: ref for ref in await n.get_references()} # Using dict instead of list
                    if len(n_refs) == 0:
                        _logger.warning("Node %s has no references, so it does not exist in Server!", ref.NodeId)
                        dangling_refs_to_missing_nodes.add(ref.NodeId)
                        continue
                    n_ref = n_refs.get(new_node_id)
                    if n_ref and n_ref.ReferenceTypeId == ref.ReferenceTypeId:
                        pass
                    else:
                        await n.add_reference(new_node_id, ref.ReferenceTypeId, not ref.IsForward)
        return dangling_refs_to_missing_nodes
```

**Reply by swamper123** [reply]
```python
                   n_ref = n_refs.get(new_node_id)
                    if n_ref and n_ref.ReferenceTypeId == ref.ReferenceTypeId:
                        pass
                    else:
                        await n.add_reference(new_node_id, ref.ReferenceTypeId, not ref.IsForward)
```


How/Where do you chek, if `new_node_id == n_ref.NodeId` ? Your dict can contain multiple references which have to be checked.
As well, a node can have different references and types. In your example, `ref` would be the last type in `await n.get_referenes()`, wouldn't it?

**Reply by xydan83** [reply]
The presence check is performed by attempting to retrieve a reference to the searched-for node from the dictionary - `n_ref = n_refs.get(new_node_id)`
Are you saying that in child nodes references (`n_refs`) can have the same **identifier** (NodeID) but different **types** (ReferenceTypeId)?

**Reply by xydan83** [reply]
The dictionary stores references of child nodes.
The key is the NodeID, the values are the ReferenceDescription, essentially a complete description of the references.
Of course, if the child nodes have references with the same NodeID, but different types of links, then this method will not work.

**Reply by xydan83** [reply]
I did several tests to find the same NodeID in the references of child nodes, until I found a situation where the identifier was repeated with different types.
Perhaps I’m not looking for that at all and didn’t understand you correctly.

**Reply by xydan83** [reply]
Compared through UAExpert reverse references that are added in the old way and through the dictionary - everything is identical.
Perhaps the data sample is small, I don’t know. 
I noticed one thing - is it really necessary to add reverse references to the standard nodes of the OPC UA types?

**Reply by swamper123** [reply]
> The presence check is performed by attempting to retrieve a reference to the searched-for node from the dictionary - n_ref = n_refs.get(new_node_id)

Sorry if I may confused you. ^^ 
I just followed your thoughts now and was confused about some things (and stuck in my old thinking of looping a list).
This seems to be a good boost!

**Reply by swamper123** [reply]
> I noticed one thing - is it really necessary to add reverse references to the standard nodes of the OPC UA types?

As far as I know:
If a new Node has a reference from the standard nodes, which is not unidirectional, the standard Node has to get a reverse reference as well.

**Reply by xydan83** [reply]
> > The presence check is performed by attempting to retrieve a reference to the searched-for node from the dictionary - n_ref = n_refs.get(new_node_id)
> 
> Sorry if I may confused you. ^^ I just followed your thoughts now and was confused about some things (and stuck in my old thinking of looping a list). This seems to be a good boost!

Perhaps at best it was possible to reduce the complexity to O (n ^ 2)
If you do not mind, how will the time be I will prepare the PR :)

**Reply by xydan83** [reply]
> > I noticed one thing - is it really necessary to add reverse references to the standard nodes of the OPC UA types?
> 
> As far as I know: If a new Node has a reference from the standard nodes, which is not unidirectional, the standard Node has to get a reverse reference as well.

Is this from the spec?

I just tried to compare data on another server, for example UAСPPServer, and there were no reverce references to demo data in the standard OPC Nodes ...

https://www.unified-automation.com/downloads/opc-ua-servers.html

It just turns out that, for example, we have added 1000 nodes and they all have at least one reference to the standard node "type of node", which means that this standard node will have 1000 reverce references. It is very cumbersome and incomprehensible why standard nodes need it?

I'll try to find it in the spec ... If you remember where this statement is located, can you send a link?

**Reply by swamper123** [reply]
> Is this from the spec?

It's how I interpreted things at that time. I remember that we had problems with importing Nodes and adding the references on them (and no complains so far). 
It is a while ago, since I tackled this topic tbh. and not great in touch with it at the moment. It is possible that some references shall not be added to the SourceNode (like unidirectional references).

If this is the case, it might be a challenge to sort out the good and the bad ones. :sweat_smile:

**Reply by xydan83** [reply]
> > Is this from the spec?
> 
> It's how I interpreted things at that time. I remember that we had problems with importing Nodes and adding the references on them (and no complains so far). It is a while ago, since I tackled this topic tbh. and not great in touch with it at the moment. It is possible that some references shall not be added to the SourceNode (like unidirectional references).
> 
> If this is the case, it might be a challenge to sort out the good and the bad ones. sweat_smile


I will try to search for these moments in the spec. It may be possible to somehow alter this search engine to further increase performance. Without this method, I did not observe any problems, it seems that all references were added, but it is quite possible that I simply have not yet encountered such a case.

**Comment by oroulet** [comment]
Is import speed an issue? I would prioritize code readability. That XML code is horrible

**Reply by xydan83** [reply]
> Is import speed an issue? I would prioritize code readability. That XML code is horrible

Rather, the problem is with one of the import methods.
